File: lib/models/book.dart
import 'package:biblioapp/models/author.dart';
import 'package:biblioapp/models/genre.dart';
import 'package:biblioapp/models/publisher.dart';

class Book {
  int id;
  String title;
  String isbn;
  int pages;
  int publicationYear;
  int editionYear;
  String synopsis;
  String coverImage;
  String pdf;
  Publisher publisher;
  List<Author> authors;
  List<Genre> genres;
  double? averageRating;

  Book({
    required this.id,
    required this.title,
    required this.isbn,
    required this.pages,
    required this.publicationYear,
    required this.editionYear,
    required this.synopsis,
    required this.coverImage,
    required this.pdf,
    required this.publisher,
    required this.authors,
    required this.genres,
    this.averageRating,
  });

  // Constructor fromJSON
  factory Book.fromJson(Map<String, dynamic> json) {
    return Book(
      id: json['id'] as int,
      title: json['title'] as String,
      isbn: json['isbn'] as String,
      pages: json['pages'] as int,
      publicationYear: json['publication_year'] as int,
      editionYear: json['edition_year'] as int,
      synopsis: json['synopsis'] as String,
      coverImage: json['cover_image'] as String,
      pdf: json['pdf'] as String,
      publisher: Publisher.fromJson(json['publisher'] as Map<String, dynamic>),
      authors:
          (json['authors'] as List)
              .map(
                (authorJson) =>
                    Author.fromJson(authorJson as Map<String, dynamic>),
              )
              .toList(),
      genres:
          (json['genres'] as List)
              .map(
                (genreJson) =>
                    Genre.fromJson(genreJson as Map<String, dynamic>),
              )
              .toList(),
      averageRating: json['reviews']?['average'] as double?,
    );
  }

  // Método toJSON
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'isbn': isbn,
      'pages': pages,
      'publication_year': publicationYear,
      'edition_year': editionYear,
      'synopsis': synopsis,
      'cover_image': coverImage,
      'pdf': pdf,
      'publisher': publisher.toJson(),
      'authors': authors.map((author) => author.toJson()).toList(),
      'genres': genres.map((genre) => genre.toJson()).toList(),
      'reviews': {'average': averageRating},
    };
  }

  // Método toString
  @override
  String toString() {
    return 'Book{id: $id, title: $title, isbn: $isbn, pages: $pages, publicationYear: $publicationYear, editionYear: $editionYear, publisher: ${publisher.name}, authors: ${authors.length}, genres: ${genres.length}}';
  }
}

File: backend_v2/models/book.rb
require_relative 'base_model'

class Book < BaseModel
  set_dataset :books
  
  many_to_one :publisher, key: :publisher_id
  one_to_many :reviews, key: :book_id
  many_to_many :authors, left_key: :book_id, right_key: :author_id, join_table: :books_authors
  many_to_many :genres, left_key: :book_id, right_key: :genre_id, join_table: :books_genres
  
  # Método de clase para obtener libros con todas las relaciones
  def self.with_all_associations
    eager(:publisher, :authors, :genres, :reviews)
  end
  
  def to_custom_json
    {
      id: id,
      title: title,
      isbn: isbn,
      pages: pages,
      publication_year: publication_year,
      edition_year: edition_year,
      synopsis: synopsis,
      cover_image: cover_image,
      pdf: pdf,
      publisher: publisher_values,
      authors: authors_values,
      genres: genres_values,
      reviews: reviews_summary
    }
  end
  
  private
  
  def publisher_values
    return nil unless publisher
    {
      id: publisher.id,
      name: publisher.name,
      logo: publisher.logo
    }
  end
  
  def authors_values
    authors.map do |author|
      {
        id: author.id,
        full_name: author.full_name,
        birth_date: author.birth_date&.strftime('%Y-%m-%d'),
        image: author.image
      }
    end
  end
  
  def genres_values
    genres.map do |genre|
      {
        id: genre.id,
        name: genre.name
      }
    end
  end
  
  def reviews_summary
    ratings = reviews.map(&:rating).compact
    average = ratings.empty? ? 0.0 : (ratings.sum.to_f / ratings.size).round(1)
    
    {
      average: average,
      count: ratings.size
    }
  end
end